File: src/dataStructures/pointIndex.ts
import { Vector } from '../dataStore';
import { chordAngFromS2Points } from '../geometry/s1/chordAngle';
import { pointFromST } from '../geometry/s2/point';
import { capFromS1ChordAngle, capGetIntersectingCells } from '../geometry/s2/cap';
import { compareIDs, convert, idFromS2Point, idRange } from '../geometry';

import type { S1ChordAngle } from '../geometry/s1/chordAngle';
import type {
  Face,
  FeatureIterator,
  JSONCollection,
  MValue,
  Projection,
  Properties,
  RGBA,
  S2CellId,
  VectorPoint,
  VectorPointM,
} from '..';
import type { VectorStore, VectorStoreConstructor } from '../dataStore';

/** The kind of input required to store a point for proper indexing */
export interface PointShape<M extends MValue = Properties> {
  cell: S2CellId;
  point: VectorPointM<M>;
}

/**
 * # Point Index
 *
 * ## Description
 * An index of cells with radius queries
 * Assumes the data is compatible with {@link https://open-s2.github.io/s2json/types/Properties.html}
 *
 * ## Usage
 * ```ts
 * import { PointIndex } from 'gis-tools-ts';
 * import { FileVector } from 'gis-tools-ts/file';
 *
 * const pointIndex = new PointIndex();
 * // or used a file based store
 * const pointIndex = new PointIndex(FileVector);
 *
 * // insert a lon-lat
 * pointIndex.insertLonLat(lon, lat, data);
 * // insert an STPoint
 * pointIndex.insertFaceST(face, s, t, data);
 *
 * // after adding data build the index. NOTE: You don't have to call this, it will be called
 * // automatically when making a query
 * await pointIndex.sort();
 *
 * // you can search a range
 * const points = await pointIndex.searchRange(low, high);
 * // or a radius
 * const points = await pointIndex.searchRadius(center, radius);
 * ```
 */
export class PointIndex<M extends MValue = Properties | RGBA> {
  #store: VectorStore<PointShape<M>>;
  #unsorted: boolean = false;

  /**
   * @param store - the store to index. May be an in memory or disk
   * @param projection - the projection of the data, defaults to S2
   */
  constructor(
    store: VectorStoreConstructor<PointShape<M>> = Vector,
    private projection: Projection = 'S2',
  ) {
    this.#store = new store();
  }

  /**
   * Set the index store to a defined one. Useful for file based stores where we want to reuse data
   * @param store - the index store
   */
  setStore(store: VectorStore<PointShape<M>>): void {
    this.#store = store;
  }

  /**
   * Insert a cell with the point and its corresponding data to the index
   * @param cell - the cell id to be indexed
   * @param point - the point to be indexed
   */
  insertID(cell: S2CellId, point: VectorPointM<M>): void {
    this.#store.push({ cell, point });
    this.#unsorted = true;
  }

  /**
   * Insert a point3D and its corresponding data to the index
   * @param point - the point to be indexed
   */
  insert(point: VectorPointM<M>): void {
    this.insertID(idFromS2Point(point), point);
  }

  /**
   * Add all points from a reader. It will try to use the M-value first, but if it doesn't exist
   * it will use the feature properties data
   * @param reader - a reader containing the input data
   */
  async insertReader(reader: FeatureIterator<unknown, M, M>): Promise<void> {
    for await (const feature of reader) this.insertFeature(feature);
  }

  /**
   * Add a vector feature. It will try to use the M-value first, but if it doesn't exist
   * it will use the feature properties data
   * @param data - any source of data like a feature collection or features themselves
   */
  insertFeature(data: JSONCollection<unknown, M, M>): void {
    const features = convert(this.projection, data, undefined, undefined, undefined, true);
    for (const { face = 0, geometry, properties } of features) {
      const { type, coordinates } = geometry;
      if (type === 'Point') {
        const { x: s, y: t, m } = coordinates;
        this.#insertFaceST(face, s, t, m ?? properties);
      } else if (type === 'MultiPoint') {
        for (const point of coordinates) {
          const { x: s, y: t, m } = point;
          this.#insertFaceST(face, s, t, m ?? properties);
        }
      }
    }
  }

  /**
   * Add a lon-lat pair to the cluster
   * @param ll - lon-lat vector point in degrees
   */
  insertLonLat(ll: VectorPoint<M>): void {
    this.insertFeature({
      type: 'VectorFeature',
      properties: ll.m ?? ({} as M),
      geometry: { type: 'Point', coordinates: ll, is3D: false },
    });
  }

  /**
   * Insert an STPoint to the index
   * @param face - the face of the cell
   * @param s - the s coordinate
   * @param t - the t coordinate
   * @param data - the data associated with the point
   */
  insertFaceST(face: Face, s: number, t: number, data: M): void {
    this.insertFeature({
      type: 'S2Feature',
      face,
      properties: data,
      geometry: { type: 'Point', coordinates: { x: s, y: t, m: data }, is3D: false },
    });
  }

  /**
   * Insert an STPoint to the index
   * @param face - the face of the cell
   * @param s - the s coordinate
   * @param t - the t coordinate
   * @param data - the data associated with the point
   */
  #insertFaceST(face: Face, s: number, t: number, data: M): void {
    this.insert(pointFromST(face, s, t, data) as VectorPointM<M>);
  }

  /**
   * iterate through the points
   * @yields a PointShape<T>
   */
  async *[Symbol.asyncIterator](): AsyncGenerator<PointShape<M>> {
    await this.sort();
    yield* this.#store;
  }

  /** Sort the index in place if unsorted */
  async sort(): Promise<void> {
    if (!this.#unsorted) return;
    await this.#store.sort();
    this.#unsorted = false;
  }

  /**
   * Find the starting index of a search
   * @param id - input id to seek the starting index of the search
   * @returns the starting index
   */
  async lowerBound(id: S2CellId): Promise<number> {
    await this.sort();
    // lower bound search
    let lo: number = 0;
    let hi: number = this.#store.length;
    let mid: number;

    while (lo < hi) {
      mid = Math.floor(lo + (hi - lo) / 2);
      const { cell: midCell } = await this.#store.get(mid);
      if (compareIDs(midCell, id) === -1) {
        lo = mid + 1;
      } else {
        hi = mid;
      }
    }

    return lo;
  }

  /**
   * Search for points given a range of low and high ids
   * @param low - the lower bound. If high is not provided, the low-high range will be created from the low
   * @param high - the upper bound
   * @param maxResults - the maximum number of results to return
   * @returns the points in the range
   */
  async searchRange(
    low: S2CellId,
    high?: S2CellId,
    maxResults = Infinity,
  ): Promise<PointShape<M>[]> {
    await this.sort();
    const res: PointShape<M>[] = [];
    if (high === undefined) {
      const [lo, hi] = idRange(low);
      low = lo;
      high = hi;
    }
    let loIdx = await this.lowerBound(low);

    while (true) {
      if (loIdx >= this.#store.length) break;
      const currLo = await this.#store.get(loIdx);
      if (compareIDs(currLo.cell, high) > 0) break;
      res.push(currLo);
      if (res.length >= maxResults) break;
      loIdx++;
    }

    return res;
  }

  /**
   * TODO: Adjust the radius for the WM projection. Really not a massive issue thogh just adjust your calcuation for now
   * Search for points within a given radius of a target point
   * @param target - the point to search
   * @param radius - the search radius
   * @param maxResults - the maximum number of results
   * @returns the points within the radius
   */
  async searchRadius(
    target: VectorPoint,
    radius: S1ChordAngle,
    maxResults = Infinity,
  ): Promise<PointShape<M>[]> {
    await this.sort();
    const res: PointShape<M>[] = [];
    if (radius < 0) return res;
    const cap = capFromS1ChordAngle<undefined>(target, radius, undefined);
    for (const cell of capGetIntersectingCells(cap)) {
      // iterate each covering s2cell min-max range on store. check distance from found
      // store Cells to target and if within radius add to results
      const [min, max] = idRange(cell);
      for (const point of await this.searchRange(min, max)) {
        if (chordAngFromS2Points(target, point.point) < radius) res.push(point);
        if (res.length >= maxResults) break;
      }
    }

    return res;
  }
}

File: rust/data_structures/point_index.rs
use crate::{
    data_store::vector::{Vector, VectorStore},
    geometry::{convert, LonLat, S1ChordAngle, S2Cap, S2CellId, S2Point},
    readers::FeatureIterator,
};
use alloc::{vec, vec::Vec};
use core::marker::PhantomData;
use s2json::{
    Face, JSONCollection, MValue, MValueCompatible, Projection, VectorFeature, VectorFeatureType,
    VectorGeometry, VectorGeometryType, VectorPoint, VectorPointGeometry,
};
use serde::{de::DeserializeOwned, Serialize};

/// # Point Index
///
/// ## Description
/// An index of cells with radius queries
/// Assumes the data is compatible with S2JSON MValues with serde_json serialization
pub struct PointIndex<
    M: MValueCompatible + Serialize + DeserializeOwned = MValue,
    S: VectorStore<VectorPoint<M>> = Vector<M>,
> {
    store: S,
    unsorted: bool,
    projection: Projection,
    _marker: PhantomData<M>,
}
impl<M: MValueCompatible + Serialize + DeserializeOwned, S: VectorStore<VectorPoint<M>>>
    PointIndex<M, S>
{
    /// Create a new PointIndex
    pub fn new(store: Option<S>, projection: Option<Projection>) -> Self {
        let store = store.unwrap_or_else(|| S::new(None));
        let projection = projection.unwrap_or(Projection::S2);
        PointIndex { store, unsorted: false, projection, _marker: PhantomData }
    }

    /// Insert a cell with the point and its corresponding data to the index
    pub fn insert(&mut self, cell: S2CellId, point: VectorPoint<M>) {
        self.store.push(cell, point);
        self.unsorted = true;
    }

    /// Insert a point3D and its corresponding data to the index. NOTE: Assumes an S2 projection.
    /// This is a quick way to add points to the index, but it is recommended to use
    /// `insert_face_st`, `insert_lon_lat`, `insert_feature`, or `insert_reader`
    pub fn insert_point(&mut self, point: VectorPoint<M>) {
        self.insert((&point).into(), point);
    }

    /// Add a lon-lat pair to the cluster
    pub fn insert_lon_lat<T: Clone>(&mut self, mut ll: LonLat<M>) {
        self.insert_feature(JSONCollection::<T, M, M>::VectorFeature(VectorFeature {
            _type: VectorFeatureType::VectorFeature,
            geometry: VectorGeometry::Point(VectorPointGeometry {
                _type: VectorGeometryType::Point,
                coordinates: ll.take(),
                is_3d: false,
                ..Default::default()
            }),
            ..Default::default()
        }));
    }

    /// Insert an STPoint to the index
    pub fn insert_face_st<T: Clone>(&mut self, face: Face, s: f64, t: f64, data: M) {
        self.insert_feature(JSONCollection::<T, M, M>::VectorFeature(VectorFeature {
            _type: VectorFeatureType::S2Feature,
            face,
            geometry: VectorGeometry::Point(VectorPointGeometry {
                _type: VectorGeometryType::Point,
                coordinates: VectorPoint::new(s, t, None, Some(data)),
                is_3d: false,
                ..Default::default()
            }),
            ..Default::default()
        }));
    }

    /// Add all points from a reader. It will try to use the M-value first, but if it doesn't exist
    /// it will use the feature properties data
    pub fn insert_reader<T: Clone, F: FeatureIterator<T, M, M>>(&mut self, reader: F) {
        for feature in reader {
            self.insert_feature(JSONCollection::<T, M, M>::VectorFeature(feature));
        }
    }

    /// Add a vector feature. It will try to use the M-value first, but if it doesn't exist
    /// it will use the feature properties data
    pub fn insert_feature<T: Clone>(&mut self, data: JSONCollection<T, M, M>) {
        let features = convert(self.projection, &data, None, None, Some(true));
        for feature in features {
            match feature.geometry {
                VectorGeometry::Point(geometry) => {
                    let coordinates = geometry.coordinates;
                    self._insert_face_st(
                        feature.face.into(),
                        coordinates.x,
                        coordinates.y,
                        coordinates.m.or(Some(feature.properties)),
                    );
                }
                VectorGeometry::MultiPoint(geometry) => {
                    for point in geometry.coordinates {
                        self._insert_face_st(
                            feature.face.into(),
                            point.x,
                            point.y,
                            point.m.or(Some(feature.properties.clone())),
                        );
                    }
                }
                _ => {}
            }
        }
    }

    /// Sort the index
    pub fn sort(&mut self) {
        if !self.unsorted {
            return;
        }
        self.store.sort();
        self.unsorted = false;
    }

    /// Find the starting index of a search
    pub fn lower_bound(&mut self, id: S2CellId) -> usize {
        self.sort();
        // lower bound search
        let mut lo: usize = 0;
        let mut hi: usize = self.store.len();
        let mut mid: usize;

        while lo < hi {
            mid = (lo + hi) / 2;
            let (mid_cell, _) = self.store.get(mid).unwrap();
            if *mid_cell < id {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }

        lo
    }

    /// Search for points given a range of low and high ids
    pub fn search_range(
        &mut self,
        mut low: S2CellId,
        high: Option<S2CellId>,
        max_results: Option<usize>,
    ) -> Vec<(S2CellId, VectorPoint<M>)> {
        let max_results = max_results.unwrap_or(usize::MAX);
        self.sort();
        let mut res = vec![];
        let high = high.unwrap_or_else(|| {
            let (lo, hi) = low.range();
            low = lo;
            hi
        });
        let mut lo_idx = self.lower_bound(low);

        loop {
            if lo_idx >= self.store.len() {
                break;
            }
            let curr_lo = self.store.get(lo_idx).unwrap();
            if curr_lo.0 > high {
                break;
            }
            res.push(curr_lo.clone());
            if res.len() >= max_results {
                break;
            }
            lo_idx += 1;
        }

        res
    }

    /// TODO: Adjust the radius for the WM projection. Really not a massive issue though just adjust your calcuation for now
    /// Search for points within a given radius of a target point
    pub fn search_radius(
        &mut self,
        target: VectorPoint,
        radius: S1ChordAngle,
        max_results: Option<usize>,
    ) -> Vec<(S2CellId, VectorPoint<M>)> {
        let max_results = max_results.unwrap_or(usize::MAX);
        self.sort();
        let mut res = vec![];
        let target: S2Point = (&target).into();
        if radius < 0. {
            return res;
        }
        let cap = S2Cap::new(target, radius, ());
        for cell in cap.get_intersecting_cells() {
            // iterate each covering s2cell min-max range on store. check distance from found
            // store Cells to target and if within radius add to results
            let (min, max) = cell.range();
            for point in self.search_range(min, Some(max), Some(max_results)) {
                if S1ChordAngle::from_s2_points(&target, &(&point.1).into()) < radius {
                    res.push(point);
                }
                if res.len() >= max_results {
                    break;
                }
            }
        }

        res
    }

    /// Insert an STPoint to the index
    fn _insert_face_st(&mut self, face: u8, s: f64, t: f64, data: Option<M>) {
        let point = S2Point::from_face_st(face, s, t);
        let vp = VectorPoint::new(point.x, point.y, Some(point.z), data);
        self.insert(point.into(), vp);
    }
}
